packages/web-api-scan-runner/src/crawl-runner/crawl-runner.spec.ts

```diff
@@ -0,0 +1,107 @@
+import { IMock, It, Mock, MockBehavior } from 'typemoq';
+import { ScanMetadataConfig } from '../scan-metadata-config';
+import { ScanMetadata } from '../types/scan-metadata';
+import { CrawlRunner } from './crawl-runner';
+
+// This class exists because trying to mock a Crawler without it
+// caused an unexplained timeout while calling the real Crawler class' constructor
+class CrawlerMock extends Crawler<string[]> {
+    constructor() {
+        super(null);
+    }
+}
+
+type CrawlerProvider = () => Promise<Crawler<string[]>>;
+
+describe('CrawlRunner', () => {
+    const scanId = 'testId';
+    const baseUrl = 'testUrl';
+    const discoveryPatterns = ['testPattern'];
+    const page = {} as Page;
+
+    let loggerMock: IMock<GlobalLogger>;
+    let scanMetaDataConfigMock: IMock<ScanMetadataConfig>;
+
+    beforeEach(() => {
+        loggerMock = Mock.ofType<GlobalLogger>(undefined, MockBehavior.Loose);
+        loggerMock.setup((m) => m.setCommonProperties({ scanId: scanId }));
+
+        scanMetaDataConfigMock = Mock.ofType<ScanMetadataConfig>(undefined, MockBehavior.Strict);
+        scanMetaDataConfigMock.setup((m) => m.getConfig()).returns(() => ({ id: scanId } as ScanMetadata));
+    });
+
+    afterEach(() => {
+        loggerMock.verifyAll();
+        scanMetaDataConfigMock.verifyAll();
+    });
+
+    it('returns undefined when crawler provider returns null', async () => {
+        const crawlerProviderMock = createCrawlerProviderMock(null);
+
+        const crawlRunner = new CrawlRunner(crawlerProviderMock.object, loggerMock.object, scanMetaDataConfigMock.object);
+
+        expect(await crawlRunner.run(baseUrl, discoveryPatterns, page)).toBeUndefined();
+
+        crawlerProviderMock.verifyAll();
+    });
+
+    it('returns undefined if crawler throws exception', async () => {
+        const crawlerMock = Mock.ofType(CrawlerMock);
+        crawlerMock.setup((m) => m.crawl(It.isAny())).throws(Error());
+
+        const crawlerProviderMock = createCrawlerProviderMock(crawlerMock.object);
+
+        const crawlRunner = new CrawlRunner(crawlerProviderMock.object, loggerMock.object, scanMetaDataConfigMock.object);
+
+        const retVal = await crawlRunner.run(baseUrl, discoveryPatterns, page);
+
+        expect(retVal).toBeUndefined();
+
+        crawlerProviderMock.verifyAll();
+        crawlerMock.verifyAll();
+    });
+
+    it('crawler receives expected data and run returns value from crawler', async () => {
+        const expectedRunOptions = {
+            baseUrl,
+            discoveryPatterns,
+            page,
+        } as SimpleCrawlerRunOptions;
+
+        let actualRunOptions: SimpleCrawlerRunOptions;
+        const expectedRetVal = ['discoveredUrl'];
+
+        const crawlerMock = Mock.ofType(CrawlerMock);
+        crawlerMock
+            .setup((m) => m.crawl(It.isAny()))
+            .callback((runOptions) => (actualRunOptions = runOptions))
+            .returns(async () => expectedRetVal);
```

Comment:
Can be simplified to
```suggestion
            .setup((m) => m.crawl(actualRunOptions))
            .returns(async () => expectedRetVal)
            .verifiable();
```
This would also remove the need to validate the value of actualRunOptions at the end of the test

Reply:
I don't think so. The parameter indicates an object to match, not an object to assign the passed in value to. Plus I tried it just to be sure and I got an error saying actualRunOptions can't be undefined, which it is at that point.

Reply:
Oops, I meant expectedRunOptions, not actualRunOptions. That will match against the expected value without the additional equality check below.

Reply:
So I didn't do this initially because I wasn't certain the equality comparison between two objects in the mock call would work. Turns out it does, so I've made the change. I will say this though: it makes the error message you get when the two objects don't match a good bit more confusing.